### ISSUE TITLE
socket / connection leak

### ISSUE THREAD

**Issue by Devhercule** [mitmproxy/mitmproxy#5135]
#### Problem Description

The mitmproxy keeps the server connection open/established while it is supposed to close it after the client (and especially the server) is manually disconnected.
The problem is particularly noticeable in the case of large file downloads.

#### Steps to reproduce the behavior:
1. start the mitmproxy : `mitmdump --set confdir=/root/MITMPYTHON/configDir/ --listen-port 85`
2. download a video with curl : `curl --proxy localhost:85 https://www.centrepompidou.fr/_elastic_filemount/media/video/77/8c/778cedf5e246fb0e0a9f0a11f2f8b114/normal.mp4  --insecure -vvv -O /dev/null`
3. monitor the status of the server connection opened by the mitmproxy with the netstat command : `netstat -anp | grep "server.sockname"` ( server.sockname is the local (ip, port) tuple for the server connection)
4. interrupt/close the client
5. again: monitor the status of the server connection `netstat -anp | grep "server.sockname"`

Here is my code (test.py) :
```
from mitmproxy import http
  
class testVideo:

  def request(self, flow: http.HTTPFlow):
    print('---------------------------- HTTPEvents : request', flow.request.url)

  def response(self, flow):
    print('---------------------------- HTTPEvents : response', flow.request.url)

  def client_connected(self, client):
    print("------ client_connected - client:(%s, %s)" % (client.peername, client.sockname))

  def client_disconnected(self, client):
    print("------ client_disconnected - client:(%s, %s)" % (client.peername, client.sockname))
  
  def server_connect(self, data):
    print("------ server_connect - client:(%s, %s) server:(%s, %s)" % (data.client.peername, data.client.sockname, data.server.sockname, data.server.peername))

  def server_connected(self, data):
    print("------ server_connected - client:(%s, %s) server:(%s, %s)" % (data.client.peername, data.client.sockname, data.server.sockname, data.server.peername))

  def server_disconnected(self, data):
    print("------ server_disconnected - client:(%s, %s) server:(%s, %s)" % (data.client.peername, data.client.sockname, data.server.sockname, data.server.peername))

  def http_connect(self, flow):
    print('---------------------------- HTTPEvents : http_connect', flow.request.url)

addons = [
  testVideo()
]
```
Config.yaml :

```
scripts:
  - "./test.py"
connection_strategy: lazy
stream_large_bodies: 30m
http2: False
websocket: False
upstream_cert: False
```
#### Current behaviour :
The server connection remains established despite the disconnection of the client and the server. 

A step by step example:

step1: start the mitmproxy and start downloading the video with curl (video size: 3240M) :

```
[root@... ~]# curl --proxy localhost:85 https://www.centrepompidou.fr/_elastic_filemount/media/video/77/8c/778cedf5e246fb0e0a9f0a11f2f8b114/normal.mp4  --insecure -vvv -O /dev/null
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0* About to connect() to proxy localhost port 85 (#0)
*   Trying 127.0.0.1...
* Connected to localhost (127.0.0.1) port 85 (#0)
* Establish HTTP proxy tunnel to www.centrepompidou.fr:443
> CONNECT www.centrepompidou.fr:443 HTTP/1.1
> Host: www.centrepompidou.fr:443
> User-Agent: curl/7.29.0
> Proxy-Connection: Keep-Alive
>
< HTTP/1.1 200 Connection established
<
* Proxy replied OK to CONNECT request
* Initializing NSS with certpath: sql:/etc/pki/nssdb
* skipping SSL peer certificate verification
* SSL connection using TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256
* Server certificate:
*       subject: CN=www.centrepompidou.fr
*       start date: Feb 13 08:54:24 2022 GMT
*       expire date: Feb 15 08:54:24 2023 GMT
*       common name: www.centrepompidou.fr
*       issuer: O=mitmproxy,CN=mitmproxy
> GET /_elastic_filemount/media/video/77/8c/778cedf5e246fb0e0a9f0a11f2f8b114/normal.mp4 HTTP/1.1
> User-Agent: curl/7.29.0
> Host: www.centrepompidou.fr
> Accept: */*
>
< HTTP/1.1 200 OK
< Server: nginx
< Date: Tue, 15 Feb 2022 07:54:24 GMT
< Content-Type: video/mp4
< Content-Length: 3397758564
< Connection: keep-alive
< Last-Modified: Sun, 18 Jul 2021 03:33:34 GMT
< ETag: "ca85ae64-5c75d7c4e56a0"
< Accept-Ranges: bytes
< Cache-Control: max-age=604800
< Expires: Tue, 22 Feb 2022 07:54:24 GMT
<
{ [data not shown]
 42 3240M   42 1379M    0     0  63.6M      0  0:00:50  0:00:21  0:00:29 68.5M
```

mitmproxy log :
```
[root@... MITMPYTHON]# mitmdump --set confdir=/root/MITMPYTHON/configDir/ --listen-port 85
Loading script ./redirect_test.py
Proxy server listening at http://*:85
------ client_connected - client:(('172.26.14.113', 52248), ('127.0.0.1', 85))
172.26.14.113:52248: client connect
---------------------------- HTTPEvents : http_connect www.centrepompidou.fr:443
---------------------------- HTTPEvents : request https://www.centrepompidou.fr/_elastic_filemount/media/video/77/8c/778cedf5e246fb0e0a9f0a11f2f8b114/normal.mp4
------ server_connect - client:(('172.26.14.113', 52248), ('127.0.0.1', 85)) server:(None, None)
172.26.14.113:52248: server connect www.centrepompidou.fr:443 (152.228.186.147:443)
------ server_connected - client:(('172.26.14.113', 52248), ('127.0.0.1', 85)) server:(('172.26.14.113', 59490), ('152.228.186.147', 443))
172.26.14.113:52248: Streaming response from www.centrepompidou.fr.
```

result of netstat :
```
[root@dlw-squid01 ~]# netstat -anp | grep 59490
tcp   3211880      0 172.26.14.113:59490     152.228.186.147:443     ESTABLISHED 15414/python
```
step2: interrupt/close the client download (ctrl+c)
mitmproxy log :
```
.............
172.26.14.113:52248: GET https://www.centrepompidou.fr/_elastic_filemount/media/video/77/8c/778cedf5e246fb0e0a9f0a1…
                  << 200 OK (content missing)
 << Client disconnected.
------ server_disconnected - client:(('172.26.14.113', 52248), ('127.0.0.1', 85)) server:(('172.26.14.113', 59490), ('152.228.186.147', 443))
------ client_disconnected - client:(('172.26.14.113', 52248), ('127.0.0.1', 85))
172.26.14.113:52248: server disconnect www.centrepompidou.fr:443 (152.228.186.147:443)
172.26.14.113:52248: client disconnect
```
=> The mitmproxy shows that the client and the server connections have been closed (server disconnect and client disconnect)
result of netstat :
```
[root@dlw-squid01 ~]# netstat -anp | grep 59490
tcp   3211880      0 172.26.14.113:59490     152.228.186.147:443     ESTABLISHED 15414/python
```
=> The mitmproxy keeps the server connexion established

**NB:** 
**The server connection is correctly closed when I let the curl command finish downloading the whole file.
The server connection remains **infinitely established** when I manually interrupt the download (the connection is established as long as mitmproxy is running).**

#### System Information
Mitmproxy: 7.0.4
Python:    3.9.5
OpenSSL:   OpenSSL 1.1.1l  24 Aug 2021
Platform:  Linux-3.10.0-1127.el7.x86_64-x86_64-with-glibc2.17

**Comment by mhils** [comment]
Thanks for the detailed report, but I can't reproduce this locally. Could you please run mitmproxy with `-vvv --set proxy_debug` and post the final lines starting from the first `ConnectionClosed`?

**Comment by Devhercule** [comment]
Here the log from the first occurrence of `ConnectionClosed`
To reproduce the problem, I suggest to let the video advance in the download (exceed 512M) before stopping it

```
172.26.14.113:47150:   >> ConnectionClosed(connection=Client({'id': '…160fa9', 'peername': ('172.26.14.113', 47150), 'sockname': ('172.26.14.113', 85), 'timestamp_start': 1644920222.0173118, 'state': <ConnectionState.CAN_WRITE: 2>, 'reply': None, 'tls': True, 'sni': 'www.centrepompidou.fr', 'alpn_offers': [], 'timestamp_tls_setup': 1644920222.2452333, 'alpn': b'', 'certificate_list': [], 'cipher': 'ECDHE-RSA-AES128-GCM-SHA256', 'tls_version': 'TLSv1.2'}))
172.26.14.113:47150:     >> ConnectionClosed(connection=Client({'id': '…160fa9', 'peername': ('172.26.14.113', 47150), 'sockname': ('172.26.14.113', 85), 'timestamp_start': 1644920222.0173118, 'state': <ConnectionState.CAN_WRITE: 2>, 'reply': None, 'tls': True, 'sni': 'www.centrep…
172.26.14.113:47150:       >> ConnectionClosed(connection=Client({'id': '…160fa9', 'peername': ('172.26.14.113', 47150), 'sockname': ('172.26.14.113', 85), 'timestamp_start': 1644920222.0173118, 'state': <ConnectionState.CAN_WRITE: 2>, 'reply': None, 'tls': True, 'sni': 'www.centrep…
172.26.14.113:47150:       << Receive(RequestEndOfMessage(stream_id=1))
172.26.14.113:47150:       >> RequestEndOfMessage(stream_id=1)
172.26.14.113:47150:         >> ConnectionClosed(connection=Client({'id': '…160fa9', 'peername': ('172.26.14.113', 47150), 'sockname': ('172.26.14.113', 85), 'timestamp_start': 1644920222.0173118, 'state': <ConnectionState.CAN_WRITE: 2>, 'reply': None, 'tls': True, 'sni': 'www.centrepompidou.fr', 'alpn_offers': [], 'timestamp_tls_setup': 1644920222.2452333, 'alpn': b'', 'certificate_list': [], 'cipher': 'ECDHE-RSA-AES128-GCM-SHA256', 'tls_version': 'TLSv1.2'}))
172.26.14.113:47150:           >> ConnectionClosed(connection=Client({'id': '…160fa9', 'peername': ('172.26.14.113', 47150), 'sockname': ('172.26.14.113', 85), 'timestamp_start': 1644920222.0173118, 'state': <ConnectionState.CAN_WRITE: 2>, 'reply': None, 'tls': True, 'sni': 'www.centrep…
172.26.14.113:47150:             >> ConnectionClosed(connection=Client({'id': '…160fa9', 'peername': ('172.26.14.113', 47150), 'sockname': ('172.26.14.113', 85), 'timestamp_start': 1644920222.0173118, 'state': <ConnectionState.CAN_WRITE: 2>, 'reply': None, 'tls': True, 'sni': 'www.centrep…
172.26.14.113:47150:               >> ConnectionClosed(connection=Client({'id': '…160fa9', 'peername': ('172.26.14.113', 47150), 'sockname': ('172.26.14.113', 85), 'timestamp_start': 1644920222.0173118, 'state': <ConnectionState.CAN_WRITE: 2>, 'reply': None, 'tls': True, 'sni': 'www.centrep…
172.26.14.113:47150:               << CloseConnection({'connection': Client({'id': '…160fa9', 'peername': ('172.26.14.113', 47150), 'sockname': ('172.26.14.113', 85), 'timestamp_start': 1644920222.0173118, 'state': <ConnectionState.CAN_WRITE: 2>, 'reply': None, 'tls': True, 'sni': 'www.centrepompidou.fr', 'alpn_offers': [], 'timestamp_tls_setup': 1644920222.2452333, 'alpn': b'', 'certificate_list': [], 'cipher': 'ECDHE-RSA-AES128-GCM-SHA256', 'tls_version': 'TLSv1.2'}), 'half_close': False})
172.26.14.113:47150:             << CloseConnection({'connection': Client({'id': '…160fa9', 'peername': ('172.26.14.113', 47150), 'sockname': ('172.26.14.113', 85), 'timestamp_start': 1644920222.0173118, 'state': <ConnectionState.CAN_WRITE: 2>, 'reply': None, 'tls': True, 'sni': 'www.cent…
172.26.14.113:47150:           << CloseConnection({'connection': Client({'id': '…160fa9', 'peername': ('172.26.14.113', 47150), 'sockname': ('172.26.14.113', 85), 'timestamp_start': 1644920222.0173118, 'state': <ConnectionState.CAN_WRITE: 2>, 'reply': None, 'tls': True, 'sni': 'www.cent…
172.26.14.113:47150:         << CloseConnection({'connection': Client({'id': '…160fa9', 'peername': ('172.26.14.113', 47150), 'sockname': ('172.26.14.113', 85), 'timestamp_start': 1644920222.0173118, 'state': <ConnectionState.CAN_WRITE: 2>, 'reply': None, 'tls': True, 'sni': 'www.cent…
172.26.14.113:47150:       << Send(ResponseProtocolError(stream_id=1, message='EOF', code=502))
172.26.14.113:47150:       >> ResponseProtocolError(stream_id=1, message='EOF', code=502)
172.26.14.113:47150:       << CloseConnection({'connection': Client({'id': '…160fa9', 'peername': ('172.26.14.113', 47150), 'sockname': ('172.26.14.113', 85), 'timestamp_start': 1644920222.0173118, 'state': <ConnectionState.CAN_WRITE: 2>, 'reply': None, 'tls': True, 'sni': 'www.centrepompidou.fr', 'alpn_offers': [], 'timestamp_tls_setup': 1644920222.2452333, 'alpn': b'', 'certificate_list': [], 'cipher': 'ECDHE-RSA-AES128-GCM-SHA256', 'tls_version': 'TLSv1.2'}), 'half_close': False})
172.26.14.113:47150:     << CloseConnection({'connection': Client({'id': '…160fa9', 'peername': ('172.26.14.113', 47150), 'sockname': ('172.26.14.113', 85), 'timestamp_start': 1644920222.0173118, 'state': <ConnectionState.CAN_WRITE: 2>, 'reply': None, 'tls': True, 'sni': 'www.cent…
172.26.14.113:47150:   << CloseConnection({'connection': Client({'id': '…160fa9', 'peername': ('172.26.14.113', 47150), 'sockname': ('172.26.14.113', 85), 'timestamp_start': 1644920222.0173118, 'state': <ConnectionState.CAN_WRITE: 2>, 'reply': None, 'tls': True, 'sni': 'www.cent…
172.26.14.113:47150:               << Receive(RequestProtocolError(stream_id=1, message='Client disconnected.', code=499))
172.26.14.113:47150:               >> RequestProtocolError(stream_id=1, message='Client disconnected.', code=499)
172.26.14.113:47150:               << Send(RequestProtocolError(stream_id=1, message='Client disconnected.', code=499))
172.26.14.113:47150:               >> RequestProtocolError(stream_id=1, message='Client disconnected.', code=499)
172.26.14.113:47150:                 >> RequestProtocolError(stream_id=1, message='Client disconnected.', code=499)
172.26.14.113:47150:                   >> RequestProtocolError(stream_id=1, message='Client disconnected.', code=499)
172.26.14.113:47150:                   << CloseConnection({'connection': Server({'id': '…3088c6', 'address': ('www.centrepompidou.fr', 443), 'state': <ConnectionState.OPEN: 3>, 'sni': 'www.centrepompidou.fr', 'tls': True, 'timestamp_start': 1644920222.2886999, 'timestamp_tcp_setup': 1644920222.2969909, 'peername': ('152.228.186.147', 443), 'sockname': ('172.26.14.113', 59510), 'alpn_offers': [], 'timestamp_tls_setup': 1644920222.3338516, 'alpn': b'', 'certificate_list': [<Cert(cn='*.centrepompidou.fr', altnames=['*.centrepompidou.fr', 'centrepom…
172.26.14.113:47150:                 << CloseConnection({'connection': Server({'id': '…3088c6', 'address': ('www.centrepompidou.fr', 443), 'state': <ConnectionState.OPEN: 3>, 'sni': 'www.centrepompidou.fr', 'tls': True, 'timestamp_start': 1644920222.2886999, 'timestamp_tcp_setup': 1644920222.…
172.26.14.113:47150:               << CloseConnection({'connection': Server({'id': '…3088c6', 'address': ('www.centrepompidou.fr', 443), 'state': <ConnectionState.OPEN: 3>, 'sni': 'www.centrepompidou.fr', 'tls': True, 'timestamp_start': 1644920222.2886999, 'timestamp_tcp_setup': 1644920222.…
172.26.14.113:47150:             << CloseConnection({'connection': Server({'id': '…3088c6', 'address': ('www.centrepompidou.fr', 443), 'state': <ConnectionState.OPEN: 3>, 'sni': 'www.centrepompidou.fr', 'tls': True, 'timestamp_start': 1644920222.2886999, 'timestamp_tcp_setup': 1644920222.…
172.26.14.113:47150:           << CloseConnection({'connection': Server({'id': '…3088c6', 'address': ('www.centrepompidou.fr', 443), 'state': <ConnectionState.OPEN: 3>, 'sni': 'www.centrepompidou.fr', 'tls': True, 'timestamp_start': 1644920222.2886999, 'timestamp_tcp_setup': 1644920222.…
172.26.14.113:47150:         << CloseConnection({'connection': Server({'id': '…3088c6', 'address': ('www.centrepompidou.fr', 443), 'state': <ConnectionState.OPEN: 3>, 'sni': 'www.centrepompidou.fr', 'tls': True, 'timestamp_start': 1644920222.2886999, 'timestamp_tcp_setup': 1644920222.…
172.26.14.113:47150:       << CloseConnection({'connection': Server({'id': '…3088c6', 'address': ('www.centrepompidou.fr', 443), 'state': <ConnectionState.OPEN: 3>, 'sni': 'www.centrepompidou.fr', 'tls': True, 'timestamp_start': 1644920222.2886999, 'timestamp_tcp_setup': 1644920222.…
172.26.14.113:47150:     << CloseConnection({'connection': Server({'id': '…3088c6', 'address': ('www.centrepompidou.fr', 443), 'state': <ConnectionState.OPEN: 3>, 'sni': 'www.centrepompidou.fr', 'tls': True, 'timestamp_start': 1644920222.2886999, 'timestamp_tcp_setup': 1644920222.…
172.26.14.113:47150:   << CloseConnection({'connection': Server({'id': '…3088c6', 'address': ('www.centrepompidou.fr', 443), 'state': <ConnectionState.OPEN: 3>, 'sni': 'www.centrepompidou.fr', 'tls': True, 'timestamp_start': 1644920222.2886999, 'timestamp_tcp_setup': 1644920222.…
172.26.14.113:47150:               << HttpErrorHook(flow=<HTTPFlow
                request = Request(GET www.centrepompidou.fr:443/_elastic_filemount/media/video/77/8c/778cedf5e246fb0e0a9f0a11f2f8b114/normal.mp4)
                response = Response(200, no content)
                error = Client disconnected.
                client_conn = Client(172.26.14.113:47150, state=closed, tls)
                server_conn = Server(www.centrepompidou.fr:443, state=closed, tls, src_port=59510)>)
172.26.14.113:47150:             << HttpErrorHook(flow=<HTTPFlow
172.26.14.113:47150:           << HttpErrorHook(flow=<HTTPFlow
172.26.14.113:47150:         << HttpErrorHook(flow=<HTTPFlow
172.26.14.113:47150:       << HttpErrorHook(flow=<HTTPFlow
172.26.14.113:47150:     << HttpErrorHook(flow=<HTTPFlow
172.26.14.113:47150:   << HttpErrorHook(flow=<HTTPFlow
172.26.14.113:47150: GET https://www.centrepompidou.fr/_elastic_filemount/media/video/77/8c/778cedf5e246fb0e0a9f0a11f2f8b114/normal.mp4
    User-Agent: curl/7.29.0
    Host: www.centrepompidou.fr
    Accept: */*
 << 200 OK (content missing)
    Server: nginx
    Date: Tue, 15 Feb 2022 10:17:02 GMT
    Content-Type: video/mp4
    Content-Length: 3397758564
    Connection: keep-alive
    Last-Modified: Sun, 18 Jul 2021 03:33:34 GMT
    ETag: "ca85ae64-5c75d7c4e56a0"
    Accept-Ranges: bytes
    Cache-Control: max-age=604800
    Expires: Tue, 22 Feb 2022 10:17:02 GMT
 << Client disconnected.
------ server_disconnected - client:(('172.26.14.113', 47150), ('172.26.14.113', 85)) server:(('172.26.14.113', 59510), ('152.228.186.147', 443))
------ client_disconnected - client:(('172.26.14.113', 47150), ('172.26.14.113', 85))
172.26.14.113:47150: server disconnect www.centrepompidou.fr:443 (152.228.186.147:443)
172.26.14.113:47150: client disconnect
172.26.14.113:47150:   >> Reply(HttpErrorHook(flow=<HTTPFlow
    request = Request(GET www.centrepompidou.fr:443/_elastic_filemount/media/video/77/8c/778cedf5e246fb0e0a9f0a11f2f8b114/normal.mp4)
    response = Response(200, no content)
    error = Client disconnected.
    client_conn = Client(172.26.14.113:47150, state=closed, tls)
    server_conn = Server(www.centrepompidou.fr:443, state=closed, tls, src_port=59510)>), None)
172.26.14.113:47150:     >> Reply(HttpErrorHook(flow=<HTTPFlow
172.26.14.113:47150:       >> Reply(HttpErrorHook(flow=<HTTPFlow
172.26.14.113:47150:         >> Reply(HttpErrorHook(flow=<HTTPFlow
172.26.14.113:47150:           >> Reply(HttpErrorHook(flow=<HTTPFlow
172.26.14.113:47150:             >> Reply(HttpErrorHook(flow=<HTTPFlow
172.26.14.113:47150:               >> Reply(HttpErrorHook(flow=<HTTPFlow
172.26.14.113:47150: closing transports...
172.26.14.113:47150: transports closed!

```

**Comment by mhils** [comment]
Thanks. I think this might be fixed with the latest development snapshots, could you check that? At least I'm seeing correct behavior here. You can find precompiled binaries at https://mitmproxy.org/downloads/#branches/main/ if that helps.

**Comment by Devhercule** [comment]
Thanks you...
The issue is fixed with `Mitmproxy: 8.0.0.dev`
The server connection is immediately closed when I interrupt the client, no more established server connection after the client and server disconnect.
Would you have an idea about the date of the new release ?

**Comment by Devhercule** [comment]
Thanks a lot
